crates/bevy_ecs/src/system/function_system.rs

```diff
@@ -108,11 +111,99 @@ impl SystemMeta {
 
     /// Marks the system as having deferred buffers like [`Commands`](`super::Commands`)
     /// This lets the scheduler insert [`apply_deferred`](`crate::prelude::apply_deferred`) systems automatically.
+    #[inline]
     pub fn set_has_deferred(&mut self) {
         self.has_deferred = true;
     }
 }
 
+impl SystemMeta {
+    /// Changes the warn policy.
+    #[inline]
+    pub fn set_param_warn_policy(&mut self, warn_policy: ParamWarnPolicy) {
+        self.param_warn_policy = warn_policy;
+    }
+
+    /// Advances the warn policy after validation failed.
+    #[inline]
+    pub fn advance_param_warn_policy(&mut self) {
+        self.param_warn_policy.advance();
+    }
+
+    /// Emits a warning about inaccessible system param if policy allows it.
+    #[inline]
+    pub fn try_warn_param<P>(&self)
+    where
+        P: SystemParam,
+    {
+        self.param_warn_policy.try_warn::<P>(&self.name);
+    }
+}
+
+/// State machine for emitting warnings when [system params are invalid](System::validate_param).
+#[derive(Clone, Copy)]
+pub enum ParamWarnPolicy {
+    /// No warning should ever be emitted.
+    Never,
+    /// The warning will be emitted once and status will update to [`Self::Never`].
+    Once,
+}
+
+impl ParamWarnPolicy {
+    /// Advances the warn policy after validation failed.
+    #[inline]
+    fn advance(&mut self) {
+        *self = match self {
+            Self::Never | Self::Once => Self::Never,
+        };
+    }
+
+    /// Emits a warning about inaccessible system param if policy allows it.
+    #[inline]
+    fn try_warn<P>(&self, name: &str)
+    where
+        P: SystemParam,
+    {
+        if matches!(self, Self::Never) {
+            return;
+        }
+        bevy_utils::tracing::warn!(
+            "System {0} will not run because it requested inaccessible system parameter {1}",
+            name,
+            core::any::type_name::<P>()
```

Comment:
Would it make sense to use `disqualified::ShortName` here?  I think that would give

```
Single<(), (With<Player>, With<Enemy>)>
```

instead of

```
bevy_ecs::system::query::Single<(), (bevy_ecs::query::filter::With<fallible_params::Player>, bevy_ecs::query::filter::With<fallible_params::Enemy>)>
```

Reply:
I'll see what I can do, never used that before

Reply:
Okay that was very easy to swap in, thanks